crates/oxc_linter/src/rules/eslint/no_empty_character_class.rs

```diff
@@ -0,0 +1,94 @@
+// Ported from https://github.com/eslint/eslint/blob/main/lib/rules/no-empty-character-class.js
+use lazy_static::lazy_static;
+use oxc_ast::AstKind;
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+use regex::Regex;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-empty-character-class): Empty character class")]
+#[diagnostic(severity(warning), help("Try to remove empty character class `[]` in regexp literal"))]
+struct NoEmptyCharacterClassDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoEmptyCharacterClass;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Disallow empty character classes in regular expressions
+    ///
+    /// ### Why is this bad?
+    /// Because empty character classes in regular expressions do not match anything, they might be typing mistakes.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// var foo = /^abc[]/;
+    /// ```
+    NoEmptyCharacterClass,
+    correctness
+);
+
+impl Rule for NoEmptyCharacterClass {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        lazy_static! {
+            /*
+            * plain-English description of the following regexp:
+            * 0. `^` fix the match at the beginning of the string
+            * 1. `([^\\[]|\\.|\[([^\\\]]|\\.)+\])*`: regexp contents; 0 or more of the following
+            * 1.0. `[^\\[]`: any character that's not a `\` or a `[` (anything but escape sequences and character classes)
+            * 1.1. `\\.`: an escape sequence
+            * 1.2. `\[([^\\\]]|\\.)+\]`: a character class that isn't empty
+            * 2. `$`: fix the match at the end of the string
+            */
+            static ref NO_EMPTY_CLASS_REGEX_PATTERN: Regex =
+                Regex::new(r"^([^\\\[]|\\.|\[([^\\\]]|\\.)+\])*$").unwrap();
```

Comment:
Is this rule just to match any []? Because it looks like literally iterating one-by-one now looking for [] next to each other that isn’t prefixed by \\ would satisfy every test now and be much clearer/faster? than the regex?

Reply:
Good point, let me check with eslint why it's written this way.

Reply:
@u9g @Boshen There is a case `var foo = /[\\[]/;` which will pass the rule, since it does not have an empty char class. So maybe we cannot simply use `contains("[]")` to test it.